src/server/snapshot.cc

```diff
@@ -37,8 +38,80 @@ size_t SliceSnapshot::DbRecord::size() const {
+    DbIndex dbid;
+    CompactObj key;
+    time_t expire;
+  };
+
+  using PendingValue = std::pair<ValueBase, util::fb2::Future<PrimeValue>>;
+  using DelayedValue = std::pair<ValueBase, PrimeValue /* stores segment and encoding flags */>;
+
+  static const int kMaxPageAccum = 100000;
+
+  void Save(DbIndex db_indx, const PrimeKey& pk, const PrimeValue& pv, time_t expire_time) {
+    ValueBase base{db_indx, PrimeKey(pk.ToString()), expire_time};
+
+    // Delay serialization of small values to possibly find more for the same page,
+    // reducing the total number of issued io requests.
+    if (auto [offset, size] = pv.GetExternalSlice(); size < TieredStorage::kMinOccupancySize) {
+      PrimeValue pv_copy;
+      pv_copy.ImportExternal(pv);
+      size_t page = offset / tiering::kPageSize;
+      auto& entries = delayed_[page];
+      {
+        delayed_sizes_.erase({entries.size(), page});
+        entries.emplace_back(std::move(base), std::move(pv_copy));
+        delayed_sizes_.insert({entries.size(), page});
+      }
+    } else {
+      pending_.push_back(Read(std::move(base), pv));
+    }
+  }
+
+  void FlushDelayed(bool force) {
+    // Flush pages with most records accumulated first, or all, if forced.
+    // It's enough just to issue reads, because they are collapsed by the tiered storage internally
+    while ((force && !delayed_.empty()) || delayed_.size() > kMaxPageAccum) {
+      DCHECK(!delayed_sizes_.empty());
+      auto [size, page] = delayed_sizes_.extract(delayed_sizes_.begin()).value();
+      auto entries = delayed_.extract(page);
+      for (auto& [base, value] : entries.mapped()) {
+        DCHECK(value.IsExternal());
+        pending_.push_back(Read(std::move(base), value));
+      }
+    }
+  }
+
+  void Serialize(RdbSerializer* serializer) {
+    for (auto& [base, value] : pending_)
+      serializer->SaveEntry(base.key, value.Get(), base.expire, base.dbid);
+    pending_.clear();
+  }
+
+ private:
+  static PendingValue Read(ValueBase value, const PrimeValue& pv) {
+    auto* ts = EngineShard::tlocal()->tiered_storage();
+    util::fb2::Future<PrimeValue> future;  // store PrimeValue directly to avoid further copies
+    ts->Read(value.dbid, value.key.ToString(), pv,
+             [future](const std::string& v) mutable { future.Resolve(PrimeValue(v)); });
+    return PendingValue{std::move(value), std::move(future)};
+  }
+
+  std::vector<PendingValue> pending_;
+
+  // Small values with delayed serialization
+  absl::flat_hash_map<size_t /* page */, std::vector<DelayedValue>> delayed_;
+  // Largest entries in delayed map
+  absl::btree_set<std::pair<size_t /* size */, size_t /* page */>, std::greater<>> delayed_sizes_;
+};
+
 SliceSnapshot::SliceSnapshot(DbSlice* slice, RecordChannel* dest, CompressionMode compression_mode)
-    : db_slice_(slice), dest_(dest), compression_mode_(compression_mode) {
+    : db_slice_(slice),
+      dest_(dest),
+      tiered_serializer_(new TieredSerializer{}),
```

Comment:
do you want to initialize it unconditionally?

Reply:
🤷🏻‍♂️ I don't think it's expensive